### DISCUSSION TITLE
A command to link backpack packages dir, what do you think?

### DISCUSSION THREAD

**Discussion by lakuapik** [Laravel-Backpack/community-forum#274] [Ideas]
When we install backpack with `php artisan backpack:install`, all files are published including the `packages` dir. The `packages` dir weight around **~87MB**. I dont really want to include all those packages into my codebase, so instead i create a command to make a symlink of `vendor/backpack/crud/src/public/packages` dir into `public/packages`. just like what `php artisan storage:link` did, now its only weight 4KB in the codebase. 

i am new with backpack, what do you think?

what is your current implementation of this?

a sneek peek of the command

```php
<?php

namespace App\Console\Commands;

use Illuminate\Console\Command;

class BackpackLink extends Command
{
    protected $signature = 'backpack:link';

    protected $description = 'Link backpack asset packages into public folder';

    public function handle(): int
    {
        $target = public_path('packages');

        if (file_exists($target)) {
            unlink($target);
        }

        $ok = symlink(
            base_path('vendor/backpack/crud/src/public/packages'),
            $target
        );

        $ok ? $this->info('Success') : $this->error('Failed');

        return $ok ? Command::SUCCESS : Command::FAILURE;
    }
}
```

**Comment by tabacitu** [comment]
I love it. Symlink is a great one-project solution.

We've been considering the symlink as a _general_ solution for a while too, but it has one big downside we can't avoid for _some_ projects: _other packages can't add stuff to the `/packages/` directory_. For example, currently:
- backpack/crud adds stuff to `packages/*`
- [backpack/filemanager](https://github.com/laravel-backpack/filemanager) adds some stuff to `packages/elfinder`

With the current directory structure & the symlink, I don't see  how you can install `backpack/filemanager` and have it running.

### Solution 1 - directory hierarchy by php packages

However, with a breaking change, we _could_:
- move the assets `backpack\crud` needs into `packages/backpack/crud`;
- move the assets `backpack\filemanager` needs into `packages/backpack/filenamanger`;

That is to say... instead of thinking of `packages` as a place where all JS & CSS assets live at the same level, we could split them by the PHP packages that need them:
```
- ...
- database/
- public/
    - assets/
    - packages/
        - backpack/
            - crud/
                - animate.css/ 
                - select2/
                - datatables/
            - filemanager/
                - elfinder/
- resources/
- ...
```


### Solution 2 - reserve 'packages' for `backpack/crud`

By convention, we could decide NOTHING gets added to `packages`, that's for `backpack/crud` only. We could even rename `packages` to  smth else, like `admin` or `backpack-crud` or `backpack-core`.

But then all addons would have to add their own CSS & JS _somewhere_, usually by publishing it, and where would they do that?


-----


Solution 1 is much cleaner I think, as it also allows people to add their own stuff to `packages`, just not in the existing directories. And it also kind of makes it obvious where those third-party libraries (datatables, jquery, select2) are used... and that you probably shouldn't edit anything there.

-----

That being said. **No matter the solution to the problem above, `symlink` method will not (and cannot) solve https://github.com/Laravel-Backpack/ideas/issues/116 (the fact that the Backpack ZIP file is big)**. Quite the opposite - it will just make it impossible to ever remove the JS&CSS assets from the PHP package. So I think it's safe to exclude `symlink` as a _general_ solution to our assets problem.

**Reply by tabacitu** [reply]
PS. Love the command, btw. Simple and effective for a lot of people. Thank you for sharing 🙏

**Reply by lakuapik** [reply]
Thank you for this comprehensive explanation. Right, this solution is only valid if we dont use any other packages/libraries that will publish to `public/packages` folder. For me, any other js/css libs would go to `public/vendor`.
